src/natcap/invest/validation.py

```diff
@@ -537,15 +564,71 @@ def check_csv(filepath, required_fields=None, excel_ok=False):
             return ("File could not be opened as a CSV. "
                     "File must be encoded as a UTF-8 CSV.")
 
-    if required_fields:
-        dataframe.columns = dataframe.columns.str.strip()
-        fields_in_table = set([name.upper() for name in dataframe.columns])
-        missing_fields = (
-            set(field.upper() for field in required_fields) - fields_in_table)
+    if header_patterns:
+
+        if axis == 1:
+            headers = [str(name).strip() for name in dataframe.iloc[0]]
+        elif axis == 0:
+            headers = [str(name).strip() for name in dataframe.iloc[:, 0]]
+        return check_patterns(header_patterns, headers)
+
+
+def check_patterns(patterns, headers):
+    """Validate that table headers (rows/columns) match a list of patterns.
+
+    - Each pattern should match at least one header.
+    - No header should be matched by more than one pattern.
+    - Each header that's matched by any pattern should not be duplicated.
+    - Headers are allowed to not match any pattern.
+    - Headers are converted to lowercase before matching.
+
+    Args:
+        patterns (list[str]): A list of patterns expected to match header(s).
+            Each pattern should be parse-able as regex and will be anchored to
+            the header start and end.
+        headers (list[str]): A list of headers to validate.
+
+    Returns:
+        None, if validation passes; or a string describing the problem, if a
+        validation rule is broken.
+    """
+    headers = [header.lower() for header in headers]  # case insensitive
+    # use a list not a dict in case there's duplicates
+    # map each expected header pattern to a list of headers that it matches
+    pattern_to_headers = [[] for pattern in patterns]
+    # map each actual header to a list of patterns that match it
+    header_to_patterns = [[] for header in headers]
+
+    for pattern_index, pattern in enumerate(patterns):
+        # add $ to the end to make it try to match the entire string
+        # re.match already anchors the pattern to the string start
+        compiled_pattern = re.compile(pattern + '$')
```

Comment:
If my pattern already includes a `$` at the end, is that a problem?

Reply:
Yeah I think the patterns should be written without `^` or `$` and assume that's handled in validation